Docs: Fundamentals of Block Development - Registration of a block (#56334)
WordPress/gutenberg: docs/getting-started/fundamentals-block-development/registration-of-a-block.md

# Registration of a block

A block is usually registered through a plugin in both the server and the client using its `block.json` metadata. 

> Although technically, blocks could be registered only on the client, **registering blocks in both the server and the client is a general recommendation**, as some features like Dynamic Rendering, Block Hooks, or Block style variations won't work properly without a server registration of the block. 

Reviewer: I would say that the one very important aspect is missing - it's the entire [Style Engine](https://github.com/WordPress/gutenberg/blob/trunk/docs/explanations/architecture/styles.md). @oandregal, what would be the way to describe it so it's obvious that registration on the server for block supports is essential for CSS styles to be generated for blocks?

Reviewer 2: I haven't reviewed everything, but the existing text sounds about right.

I don't know how deep you want to go into this, or if you rather link to existing documents. What I could add is that "global styles" (styles generated from `theme.json`) require blocks to be registered in the server – otherwise, it's like they don't exist, and any styles assigned to them in `theme.json` will be ignored. This only matters if the block aims to be styled via `theme.json`, of course.

I presume the same would be true for all features that generate styles in the server (block supports, layout, style engine, etc.). @aaronrobertshaw @ramonjd @andrewserong @tellthemachines have implemented most of it. They know more than I do on this topic!

Reviewer: Excellent feedback André. Much appreciate inviting other implementers to share more context 🙇🏻

Author: Thanks @gziolo @oandregal for the feedback! 
I've updated the PR taking this feedback into account. 
Let me know your thoughts.

> I don't know how deep you want to go into this, or if you rather link to existing documents.

The purpose of this "Getting Started > Fundamentals of Block Development" is to introduce key ideas to help developers have a better "onboarding" experience for block development. Links to other parts of the Handbook will be used for more details, but in some specific cases, the content will be directly "moved" to this section.

Reviewer 3: Hi folks, thanks for the ping, and great work on the docs. 

In relation to the style engine, along with the general [README/usage](https://github.com/WordPress/gutenberg/blob/trunk/packages/style-engine/README.md) docs there is some block-supports-specific documentation written here:

https://github.com/WordPress/gutenberg/blob/trunk/packages/style-engine/docs/using-the-style-engine-with-block-supports.md

And there's the handbook entry for [block supports](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/).

As @juanmaguitar suggests, maybe it's enough to refer and link to the relevant docs as appropriate should folks wish to enable block supports features to their newly-registered block?

P.S., Going over this stuff made me realize I need to update the style engine block supports docs so I have a [PR](https://github.com/WordPress/gutenberg/pull/56528) to do that 😆 So thank you!

# Registration of a block

> A block is usually registered through a plugin in both the server and the client using its `block.json` metadata. 

Reviewer 3: ```suggestion
A block is usually registered through a plugin on both the server and client-side using its `block.json` metadata. 
```

"On" the server I think the most usual preposition for server and "in" when talking about a browser client. I don't know why 😄 

Feel free to ignore this suggestion, as the meaning is clear regardless. 😄

> Although technically, blocks could be registered only on the client, **registering blocks in both the server and the client is a strong recommendation**, as some server-side features like Dynamic Rendering, Block Supports, Block Hooks, or Block style variations require the block to "exist" in the server and they won't work properly without a server registration of the block.

Reviewer 3: ```suggestion
Although technically, blocks could be registered only in the client, **registering blocks on both the server and in the client is a strong recommendation**. Some server-side features like Dynamic Rendering, Block Supports, Block Hooks, or Block style variations require the block to "exist" on the server, and they won't work properly without server registration of the block.
```

> For example, to enable the [capability of blocks to be styled via `theme.json`](https://developer.wordpress.org/themes/global-settings-and-styles/settings/blocks/), they need to be registered in the server, otherwise, any styles assigned to them in `theme.json` will be ignored. 

Reviewer 3: ```suggestion
For example, to allow a block [to be styled via `theme.json`](https://developer.wordpress.org/themes/global-settings-and-styles/settings/blocks/), it needs to be registered on the server, otherwise, any styles assigned to it in `theme.json` will be ignored. 
```

[![Open Block Registration diagram in excalidraw](https://developer.wordpress.org/files/2023/11/block-registration-e1700493399839.png)](https://excalidraw.com/#json=PUQu7jpvbKsUHYfpHWn7s,61QnhpZtjykp3s44lbUN_g "Open Block Registration diagram in excalidraw")

### Registration of the block with PHP (server-side)

> A block registration in the server usually takes place in the main plugin PHP file with the `register_block_type` function called on the `init` action.

Reviewer 3: ```suggestion
Block registration on the server usually takes place in the main plugin PHP file with the `register_block_type` function called on the [init hook](https://developer.wordpress.org/reference/hooks/init/).
```

> The [`register_block_type`](https://developer.wordpress.org/reference/functions/register_block_type/) function, that aims to simplify the block type registration on the server, can read metadata stored in the `block.json` file.

Reviewer 3: ```suggestion
The [`register_block_type`](https://developer.wordpress.org/reference/functions/register_block_type/) function aims to simplify block type registration on the server by reading metadata stored in the `block.json` file.
```

This function takes two params relevant in this context (`$block_type` accepts more types and variants):

-   `$block_type` (`string`) – path to the folder where the `block.json` file is located or full path to the metadata file if named differently.
-   `$args` (`array`) – an optional array of block type arguments. Default value: `[]`. Any arguments may be defined. However, the one described below is supported by default:
    -   `$render_callback` (`callable`) – callback used to render blocks of this block type, it's an alternative to the `render` field in `block.json`.

> As part of the build process, the `block.json` is usually copied from the `src` folder to the `build` folder, so the `block.json` registered should be the one in the `build` folder.

Reviewer 3: ```suggestion
As part of the build process, the `block.json` file is usually copied from the `src` folder to the `build` folder, so the path to the `block.json` of your registered block should refer to the `build` folder.
```

The function takes two params:

> -   `$blockNameOrMetadata` (`string`|`Object`) – block type name (if previously registered on the server) or the metadata object loaded from the `block.json` file with a bundler (e.g., webpack) or a custom Babel plugin.

Reviewer: The block can be registered only on the client, so I'm not sure whether the rephrased part fits in that case. Otherwise, it's valid.

Author: The text in parenthesis `(if previously registered on the server)` meant to clarify that registering a block in the client only using `block type name` only works when the block's metadata has been previously registered on the server.

I'm not sure what `(supported previously)` is referring to.

But I think I can remove this clarification as this is already explained at the beginning of the `Registration of the block with JavaScript (client-side)` section

> When the block is registered on the server, you only need to register the client-side settings on the client using the same block’s name.
